### Analyze the data on  Gun Deaths in the US 
In this project, we analyze gun detahs in the US. The dataset came from FiveThirtyEight. The dataset is stored in the guns.csv file from 2012 to 2014.

In [1]:
import csv
import datetime 
file = open('guns.csv', 'r')

data = list(csv.reader(file, delimiter=','))
data[0:5]

[['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 ['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]

In [2]:
#extract header row, then remove header row from the list
headers = data[0]
print(headers)
data = data[1:]
data[0:5]

['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']


[['1',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '34',
  'Asian/Pacific Islander',
  '100',
  'Home',
  '4'],
 ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'],
 ['3',
  '2012',
  '01',
  'Suicide',
  '0',
  'M',
  '60',
  'White',
  '100',
  'Other specified',
  '4'],
 ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
 ['5',
  '2012',
  '02',
  'Suicide',
  '0',
  'M',
  '31',
  'White',
  '100',
  'Other specified',
  '2']]

The year column contains information on the year in which the gun deaths occured. We can use this column to calculate how many gun deaths happened in each year.
We can perform this operation by creating a dictionary, then keeping count in the dictionary of how many times each element occurs in the year column.

In [3]:
years = [row[1] for row in data]

year_counts = dict()

for year in years:
    if year in year_counts:
        year_counts[year] +=  1
    else:
        year_counts[year] =  1
print("Number of gun deaths by year from 2012 to 2014: ")        
print("Year                 Gun Deaths")
print("-------------------------------")
for k,v in year_counts.items():
    print(k, "                 ", v)


Number of gun deaths by year from 2012 to 2014: 
Year                 Gun Deaths
-------------------------------
2013                   33636
2012                   33563
2014                   33599


In [4]:
dates = [datetime.datetime(year=int(row[1]), month=int(row[2]), day=1 ) for row in data]


In [5]:
date_counts = dict()
for date in dates:
    if date in date_counts:
        date_counts[date] +=1
    else:
        date_counts[date] =1


In [6]:
# sex column = 5 and race column = 7 where index starts at 0
sex_counts = dict()
race_counts = dict()
for row in data:
    sex = row[5]
    if sex in sex_counts:
        sex_counts[sex] += 1
    else:
        sex_counts[sex] = 1
    race = row[7]
    if race in race_counts:
        race_counts[race] += 1
    else:
        race_counts[race] = 1
print("Gun deaths by Sex: ")
print("---------------------")
print("Sex","   ", "Gun Deaths")
print("---------------------")
for k,v in sex_counts.items():
    print(k, "         ", v)

print("\nGun deaths by Race: ")
print("--------------------------------------------------")
print("Race", "                         ", "Gun Deaths")
print("--------------------------------------------------")
space = 32
for k,v in race_counts.items():
    strlen = 33 - len(k)
    spacestr = ""
    for i in range(1,strlen):
        spacestr += " "
    print(k, spacestr, v)


Gun deaths by Sex: 
---------------------
Sex     Gun Deaths
---------------------
M           86349
F           14449

Gun deaths by Race: 
--------------------------------------------------
Race                           Gun Deaths
--------------------------------------------------
Native American/Native Alaskan    917
Hispanic                          9022
White                             66237
Asian/Pacific Islander            1326
Black                             23296


We explored gun deaths by race in the last cell. However, our analysis only gives us the total number of gun deaths by race in the US. Unless we know the proportion of each race in the US, we won't be able to meaningfully compare the numbers. 
What we really want to get is a rate of gun deaths per 100000 people of each race. In order to do this, we'll nedd to read in data about what percentage of the US population falls into each racial category. We can import census data to explore the propertion of gun deaths.

The census.csv file contains infromation on the total population of the US, as well as the total population of each racial group in the US. 

In [7]:
censusFile = open("census.csv", "r")
census = list(csv.reader(censusFile, delimiter=","))
census

[['Id',
  'Year',
  'Id',
  'Sex',
  'Id',
  'Hispanic Origin',
  'Id',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 ['cen42010',
  'April 1, 2010 Census',
  'totsex',
  'Both Sexes',
  'tothisp',
  'Total',
  '0100000US',
  '',
  'United States',
  '308745538',
  '197318956',
  '44618105',
  '40250635',
  '3739506',
  '15159516',
  '674625',
  '6984195']]

#### Calculating gun deaths per 100K of the population by race
Earlier, we computed the number of gun deaths per race, and created a dictionary, race_counts. In order to get from the raw counts of gun deaths by race to a rate of gun deaths per 100000 people in each race, we'll need to divide the total number of gun deaths by the population of each race. 
So for example the total number of white people killed from the period of 2012 to 2014 is 66237. From the census data the total white population was 197318956. So the rate of white people killed per 100000 is :
lets say N is the total white population from the census data
and K is the total gun deaths among the white population from year 2012 to 2014.

white_gun_death_rate = $\frac{K}{N}$

white_gun_death_rate = $66237/197318956$ = 0.0003356849303419181

rate_per_hundredk = 0.0003356849303419181 * 100000 = 33.56


This gives us 33.56, which can be interpreted as 33.56 out of every 100000 people in the White census race category in the US were killed by guns between 2012 and 2014.

We will need to calculate the same rates for each racial category. The challenge is that the racial categories are named slightly different in census and in data. We''ll need to manually construct a dictionary that allows us to map between them and perform the division.

In [8]:
mapping = dict()
mapping["Asian/Pacific Islander"] = int(census[1][14])+int(census[1][15])
mapping["Black"] = int(census[1][12])
mapping["Native American/Native Alaskan"] = int(census[1][13])
mapping["Hispanic"] = int(census[1][11])
mapping["White"] = int(census[1][10])
mapping

{'Asian/Pacific Islander': 15834141,
 'Black': 40250635,
 'Hispanic': 44618105,
 'Native American/Native Alaskan': 3739506,
 'White': 197318956}

In [9]:
race_per_hundredk = dict()
for k,v in race_counts.items():
    race_per_hundredk[k] = (v/mapping[k])*100000
race_per_hundredk

{'Asian/Pacific Islander': 8.374309664161762,
 'Black': 57.8773477735196,
 'Hispanic': 20.220491210910907,
 'Native American/Native Alaskan': 24.521955573811088,
 'White': 33.56849303419181}

#### Filtering by intent ####
We can filter the results, and restrict them to the "Homicide" intent. This will tell us what the gun-related murder rate per 100000 people is in each racial category. In order to do this we will have to recompute the race_counts dictionary but only count rows where the intent is "Homicide".

In [10]:
intents = [row[3] for row in data]
races = [row[7] for row in data]
homicide_race_per_hundredk = dict()
for i, race in enumerate(races):
    if intents[i].lower() == 'homicide':
        if race in homicide_race_per_hundredk:
            homicide_race_per_hundredk[race] += 1
        else:
            homicide_race_per_hundredk[race] = 1
for k,v in homicide_race_per_hundredk.items():
        homicide_race_per_hundredk[k] = (v/mapping[k])*100000
print("The rate of homicide per 100,000 people by race:")        
homicide_race_per_hundredk

The rate of homicide per 100,000 people by race:


{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}